src/Cryptol/TypeCheck/Unify.hs

```diff
@@ -111,21 +111,17 @@ bindVar v@(TVBound {}) t
   where k = kindOf v
 
 bindVar x@(TVFree _ _ xscope _) (TVar y@(TVFree _ _ yscope _))
-  | xscope `Set.isProperSubsetOf` yscope = return (singleSubst y (TVar x), [])
+  | xscope `Set.isProperSubsetOf` yscope =
+    return (uncheckedSingleSubst y (TVar x), [])
```

Comment:
It looks like this case is redundant, as the next one already handles it?

Reply:
This case does not overlap with the next case: It is for situations where we can add the mapping `y -> x` to the substitution, but instantiating `x -> y` would be forbidden because `y` would escape from its scope.

It seems like a comment explaining this might be in order.

Reply:
I just realized that we should probably have an explicit kind check in this case, like we do in the following one.

Reply:
Good point, I didn't see that we'd swapped that `x` and `y`.